tests: lib: cmsis:transform: cf64 min_ram 128k

ABOSTM opened the pull request:
tests: lib: cmsis:transform: cf64 min_ram 128k

tests/lib/cmsis_dsp/transform/ with config:
   * libraries.cmsis_dsp.transform.cf64
Needs more than 64k RAM.
128k is sufficient if tests steps are reordered
(decreasing malloc size)

Fixes #27849

ABOSTM committed:
tests: lib: cmsis:transform: cf64 min_ram 96k

tests/lib/cmsis_dsp/transform/ with config:
   * libraries.cmsis_dsp.transform.cf64
Needs more than 64k RAM.

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

stephanosio commented on tests/lib/cmsis_dsp/transform/testcase.yaml:
```diff
@@ -47,7 +47,7 @@ tests:
   libraries.cmsis_dsp.transform.cf64:
     min_flash: 1024
-    min_ram: 64
+    min_ram: 128
```
> But while executing step test_arm_cfft_f64_cmplx_noisy_4096 there is a malloc() of length:
> 8192 * sizeof(float64_t) = 8192 * 8 = 64k
> but we need more RAM for variables, stack, ...

As you noted in #27849, `min_ram: 64` is indeed insufficient.

I am not sure if setting this to `128` is the best solution though; `80` or `96` should (at least, theoretically) work.

ABOSTM replied:
I dig a bit a found a nucleo_l476rg (96k ram 1024 flash) able to pass the test.
And I found no board (any manufacturers) with 80k ram and enough flash (1024)
So I propose to set min_ram: 96

ABOSTM replied:
Split done
set 96 done

stephanosio commented on tests/lib/cmsis_dsp/transform/src/cf64.c:
```diff
@@ -167,42 +167,42 @@ DEFINE_TEST_VARIANT4(arm_cifft_f64_cmplx, step_4096, &arm_cfft_sR_f64_len4096,
-		ztest_unit_test(test_arm_cifft_f64_cmplx_step_512),
-		ztest_unit_test(test_arm_cfft_f64_cmplx_step_1024),
-		ztest_unit_test(test_arm_cifft_f64_cmplx_step_1024),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_noisy_2048),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_noisy_2048),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_noisy_1024),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_noisy_1024),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_noisy_512),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_noisy_512),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_noisy_256),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_noisy_256),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_noisy_128),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_noisy_128),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_noisy_64),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_noisy_64),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_noisy_32),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_noisy_32),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_noisy_16),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_noisy_16),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_4096),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_4096),
 		ztest_unit_test(test_arm_cfft_f64_cmplx_step_2048),
 		ztest_unit_test(test_arm_cifft_f64_cmplx_step_2048),
-		ztest_unit_test(test_arm_cfft_f64_cmplx_step_4096),
-		ztest_unit_test(test_arm_cifft_f64_cmplx_step_4096)
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_1024),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_1024),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_512),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_512),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_256),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_256),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_128),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_128),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_64),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_64),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_32),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_32),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_16),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_16)
```
From #27849, re:
> I don't know exactly how malloc() algorithm from newlibc is working,
> but it appears during my debug session that even if memory is freed between each step,
> each time the size of the malloc is increased, the previous memory size is 'lost' (probably due to heap fragmentation).

IMO, this needs more investigation. I am not sure if fragmentation is the cause as far as this test goes (there aren't all that many mixed-size mallocs here to begin with -- all the allocated blocks are freed before subsequent different-size mallocs); rather, it seems there is something strange going on.

ABOSTM replied:
IMO fragmentation is the issue:
When test was in increasing order, the return of malloc() was respectively:
0x200017A8
0x200018B0
0x20001AB8
0x20001EC0
0x200026C8
0x200036D0 (step 512, malloc 0x2000)
0x200056D8 (step 1024, malloc 0x4000)
0x200096E0  (step 2048, malloc 0x8000)
Each time the malloc pointer is a bit more than the previous malloc.
I can just assume maaloc algo can reuse freed chunck for smaller malloc but cannot reuse for lager one.

ioannisg replied:
@stephanosio ^^

stephanosio replied:
Fair enough.

stephanosio reviewed (commented):
Could you split the changes in `testcase.yml` and `cf64.c` into separate commits?

Ideally, updating the memory requirement in `testcase.yml` should be sufficient; but, it seems an additional change (reordering) is required due to something funny going on inside newlib malloc -- which ought to be investigated and fixed; after which, it should no longer be necessary and therefore reverted.

ABOSTM committed:
tests: lib: cmsis:transform: cf64 min_ram reorder test to fit 96k

tests/lib/cmsis_dsp/transform/ with config:
   * libraries.cmsis_dsp.transform.cf64
96k is sufficient if tests steps are reordered
(decreasing malloc size)
Tested on nucleo_l476rg

Signed-off-by: Alexandre Bourdiol <alexandre.bourdiol@st.com>

ABOSTM commented:
Similar PR for tests/lib/cmsis_dsp/matrix/: #27925

carlescufi commented:
@erwango @ioannisg please review

ioannisg reviewed (commented):
Even if the fragmentation is not the root cause if this problem (which this PR is trying to fix), I think this PR is fine as it is, theoretically, a "no-op" : it only re-orders the tests.

ioannisg commented on tests/lib/cmsis_dsp/transform/testcase.yaml:
```diff
@@ -47,7 +47,7 @@ tests:
   libraries.cmsis_dsp.transform.cf64:
     min_flash: 1024
-    min_ram: 64
+    min_ram: 96
```
@ABOSTM  Does this really need more than 64K of RAM in several ones or only this particular platform?
Could you try to build in some more ARM Boards with 64K RAM and see what is going on?

ABOSTM replied:
@ioannisg , as explained in https://github.com/zephyrproject-rtos/zephyr/issues/27849
there is a malloc of 64K so when you add stack, variables ... 
So for sure, 64K is not enough.

ioannisg replied:
agreed, thanks

ioannisg commented:
Not urgent to be merged before the 2.4 freeze - this is probably fixing a bug, so it can be dealt with later.
